### PR TITLE
Add a native SwiftUI button

### PR DESCRIPTION
**Pull request by mdmathias**
This change adds a native SwiftUI button, and seeks to provide further native Swift support (#49).

This PR will also lay the foundation for more Swift-specific development by establishing a separate library.

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 import PackageDescription
 
-let googleSignInVersion = "6.1.0"
+let googleSignInVersion = "6.2"
```
Comment (petea) [inline Package.swift:20]: Can we wait on the version bump?  I usually do this all at once in a separate PR.

Comment (mdmathias) [reply Package.swift:20]: Sure thing.

File: GoogleSignIn/Sources/GIDSignInButtonStyling.m [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+const NSUInteger kNumGIDSignInButtonStyleColors = 2;
+
+@interface GIDSignInButtonStyling ()
+
+@end
+
+@implementation GIDSignInButtonStyling
+
++ (NSURL *)googleIconURL {
+  NSURL *imgURL = [SWIFTPM_MODULE_BUNDLE URLForResource:kGoogleImageName
```
[the remark this answers is not in the document]
Comment (mdmathias) [reply GoogleSignIn/Sources/GIDSignInButtonStyling.m]: Good point. Fixed.

**Comment by petea** [comment]
Look into using the --include-podspecs flag for pod lib lint for loading the GoogleSignIn podspec from the local repo.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -15,19 +15,25 @@ jobs:
...
-          '--use-static-frameworks'
+        podspec: [GoogleSignIn.podspec, GoogleSignInSwift.podspec]
+        flag: [
+          "", 
+          "--use-libraries", 
+          "--use-static-frameworks"
         ]
+        include:
+          - podspec: GoogleSignInSwift.podspec
+            includePodspecFlag: "--include-podspecs='*.podspec'"
```
Comment (petea) [inline .github/workflows/tests.yml:26]: Now that we're only targeting the Swift pod run, we could be more specific:

`--include-podspecs='GoogleSignIn.podspec'`

File: GoogleSignInSwift.podspec [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  s.osx.deployment_target = macos_deployment_target
+  s.prefix_header_file = false
+  s.source_files = [
+    'GoogleSignInSwift/Sources/*.swift',
+  ]
+  s.frameworks = [
+    'CoreGraphics',
+    'SwiftUI',
+  ]
+  s.dependency 'GoogleSignIn', '~> 6.1'
```
Comment (petea) [inline GoogleSignInSwift.podspec:28]: Update this to 6.2 after rebasing from main to bring in my 6.2 version bump.

File: GoogleSignInSwift.podspec [hunk 5]
```diff
@@ -0,0 +1,29 @@
+Pod::Spec.new do |s|
+  s.name = 'GoogleSignInSwift'
+  s.version = '0.1.0'
```
Comment (petea) [inline GoogleSignInSwift.podspec:3]: I think we should sync versions with the main pod, setting this to 6.2.0.  Although we could have independent versioning for the pods, we couldn't do this for the SPM products because SPM uses one git tag to version all products in the package definition.

File: GoogleSignInSwift.podspec [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  s.swift_version = '4.0'
+  s.summary = 'Adds Swift-focused support for Google Sign In.'
+  s.description = <<-DESC
+The Google Sign-In SDK allows users to sign in with their Google account from third-party apps.
```
Comment (petea) [inline GoogleSignInSwift.podspec:7]: Let's just sync this text with the summary for now.

File: GoogleSignIn/Sources/GIDSignInButtonStyling.m [hunk 7]
```diff
@@ -0,0 +1,120 @@
+// Copyright 2021 Google LLC
```
Comment (petea) [inline GoogleSignIn/Sources/GIDSignInButtonStyling.m:1]: "2022" here and in the other new class files.

File: GoogleSignIn/Sources/GIDSignInButtonStyling.m [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButtonStyling.h"
+#import "GoogleSignIn/Sources/NSBundle+GID3PAdditions.h"
+
```
Comment (petea) [inline GoogleSignIn/Sources/GIDSignInButtonStyling.m:17]: NS_ASSUME_NONNULL_BEGIN ?

File: GoogleSignIn/Sources/GIDSignInButtonStyling.m [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  return [NSURL URLWithString:iconPath];
+}
+
+/// Retrieves the Google font URL from the bundle.
++ (NSURL *)googleFontURL {
+  NSBundle *bundle = [NSBundle gid_frameworkBundle];
+  NSString *fontPath = [bundle pathForResource:kFontNameRobotoBold ofType:@"ttf"];
+  return [NSURL URLWithString:fontPath];
+}
+
```
Comment (petea) [inline GoogleSignIn/Sources/GIDSignInButtonStyling.m:119]: NS_ASSUME_NONNULL_END ?

File: GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButton.h [hunk 10]
```diff
@@ -19,26 +19,11 @@
...
+// Forward declarations of enums
+typedef NS_ENUM(NSInteger, GIDSignInButtonStyle);
+typedef NS_ENUM(NSInteger, GIDSignInButtonColorScheme);
 
-/// The color schemes supported by the `GIDSignInButton`.
-typedef NS_ENUM(NSInteger, GIDSignInButtonColorScheme) {
-  kGIDSignInButtonColorSchemeDark = 0,
-  kGIDSignInButtonColorSchemeLight = 1
-};
+NS_ASSUME_NONNULL_BEGIN
```
Comment (petea) [inline GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButton.h:26]: Move this back above typedefs for consistency?

File: GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButton.h [hunk 11]
```diff
@@ -19,26 +19,11 @@
...
-/// The following dimensions (in points) fit for all languages:
-/// - kGIDSignInButtonStyleStandard: 230 x 48
-/// - kGIDSignInButtonStyleWide:     312 x 48
-/// - kGIDSignInButtonStyleIconOnly: 48 x 48 (no text, fixed size)
-typedef NS_ENUM(NSInteger, GIDSignInButtonStyle) {
-  kGIDSignInButtonStyleStandard = 0,
-  kGIDSignInButtonStyleWide = 1,
-  kGIDSignInButtonStyleIconOnly = 2
-};
+// Forward declarations of enums
```
Comment (petea) [inline GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButton.h:22]: If we don't define the enum values here, is this not a breaking change?

**Review by petea — CHANGES_REQUESTED** [review]

File: GoogleSignInSwift.podspec (hunk 4 above)
Comment (mdmathias) [reply GoogleSignInSwift.podspec:28]: Sounds good.

File: GoogleSignInSwift.podspec (hunk 5 above)
Comment (mdmathias) [reply GoogleSignInSwift.podspec:3]: Okay, I'll set this to 6.2.0. One great leap forward!

File: GoogleSignInSwift.podspec (hunk 6 above)
Comment (mdmathias) [reply GoogleSignInSwift.podspec:7]: Done

File: GoogleSignInSwift.podspec [hunk 12]
```diff
@@ -0,0 +1,27 @@
+Pod::Spec.new do |s|
+  s.name = 'GoogleSignInSwift'
+  s.version = '6.2.0'
+  s.swift_version = '4.0'
+  s.summary = 'Adds Swift-focused support for Google Sign In.'
+  s.description = 'Adds Swift-focused support for Google Sign In.'
```
Comment (petea) [inline GoogleSignInSwift.podspec:6]: Looks like I broke your pod lib lint here.  How about we try:

"Additional Swift support for the Google Sign-In SDK."

File: GoogleSignInSwift.podspec [hunk 13]
```diff
@@ -0,0 +1,27 @@
+Pod::Spec.new do |s|
+  s.name = 'GoogleSignInSwift'
+  s.version = '6.2.0'
+  s.swift_version = '4.0'
+  s.summary = 'Adds Swift-focused support for Google Sign In.'
```
Comment (petea) [inline GoogleSignInSwift.podspec:5]: Don't forget the hyphen: "Google Sign-In"

File: GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButton.h (hunk 11 above)
Comment (mdmathias) [reply GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButton.h:25]: I've imported `GIDSignInButtonStyling.h` into `GIDSignInButton.h`  to circumvent this. Per offline chat, I agree that this isn't desirable as we're leaking implementation details into our API, but I cannot see a way of creating a Swift target that depends upon this information without doing something like that. We can treat this as a temporary solution and create an issue to track fixing this.

**Review by guilhermedfs — APPROVED** [review]

File: GoogleSignInSwift/Sources/GIDSignInButton_SwiftUI.swift [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+    switch self {
+    case .iconOnly: return Width(min: CGFloat(kIconWidth), max: CGFloat(kIconWidth))
+    case .standard: return Width(min: 90, max: .infinity)
+    case .wide: return Width(min: 170, max: .infinity)
+    default:
+      fatalError("Unrecognized case for `GIDSignInButtonStyle: \(self)")
+    }
+  }
+
+  fileprivate var buttonText: String {
```
Comment (petea) [inline GoogleSignInSwift/Sources/GIDSignInButton_SwiftUI.swift:44]: We'll need to load strings from the shared bundle in order to handle localization.

File: GoogleSignInSwift/Sources/GIDSignInButton_SwiftUI.swift [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+struct SwiftUIButtonStyle: ButtonStyle {
+  let style: GIDSignInButtonStyle
+
+  func makeBody(configuration: Configuration) -> some View {
+    configuration.label
+      .frame(minWidth: style.width.min,
+             maxWidth: style.width.max,
+             minHeight: CGFloat(kButtonHeight),
+             maxHeight: CGFloat(kButtonHeight))
+      .foregroundColor(.blue)
```
Comment (petea) [inline GoogleSignInSwift/Sources/GIDSignInButton_SwiftUI.swift:66]: The default text color for the SwiftUI button seems to be blue rather than grey as in GIDSignInButton, is this the cause?

**Review by petea — CHANGES_REQUESTED** [review]

File: GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButton.h (hunk 11 above)
Comment (petea) [reply GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButton.h:25]: Discussed offline and agreed to pursue a refactor with the goal of not having any public API surface impact.

File: Samples/Swift/DaysUntilBirthday/DaysUntilBirthday.xcodeproj/project.pbxproj [hunk 16]
```diff
@@ -384,22 +389,27 @@
...
 /* Begin XCRemoteSwiftPackageReference section */
-		7345AD252703DA3E0020AFB1 /* XCRemoteSwiftPackageReference "GoogleSignIn-iOS" */ = {
+		73F87DE427C9B98A00E12BFE /* XCRemoteSwiftPackageReference "GoogleSignIn-iOS" */ = {
 			isa = XCRemoteSwiftPackageReference;
-			repositoryURL = "https://github.com/google/GoogleSignIn-iOS.git";
+			repositoryURL = "https://github.com/google/GoogleSignIn-iOS";
 			requirement = {
-				kind = upToNextMajorVersion;
-				minimumVersion = 6.0.2;
+				branch = "mdmathias/swiftui-button";
```
[the remark this answers is not in the document]
Comment (mdmathias) [reply Samples/Swift/DaysUntilBirthday/DaysUntilBirthday.xcodeproj/project.pbxproj:396]: I'll remove it before submitting.

File: GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButtonStyling.h [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+// Button text font size.
+extern const CGFloat kFontSize;
+
+#pragma mark - Icon Constants
+
+// The name of the image for the Google "G"
+extern NSString *const kGoogleImageName;
+
+// Keys used for NSCoding.
```
[the remark this answers is not in the document]
Comment (mdmathias) [reply GoogleSignIn/Sources/Public/GoogleSignIn/GIDSignInButtonStyling.h]: Removed the file.

File: Samples/Swift/DaysUntilBirthday/DaysUntilBirthday.xcodeproj/project.pbxproj (hunk 16 above)
[the remark this answers is not in the document]
Comment (petea) [reply Samples/Swift/DaysUntilBirthday/DaysUntilBirthday.xcodeproj/project.pbxproj:396]: Pin's recently merged #111 adds GoogleSignIn as a local package, so you should be able to simply remove this after merging.

[the remark this answers is not in the document]
Comment (petea) [reply Samples/Swift/DaysUntilBirthday/DaysUntilBirthday.xcodeproj/project.pbxproj:396]: Pin's recently merged #111 adds GoogleSignIn as a local package, so you should be able to simply remove this after merging.

File: GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+ */
+
+import SwiftUI
+import CoreGraphics
+
+// MARK: - Sign-In Button
+
+/// A Google Sign In button to be used in SwiftUI.
+@available(iOS 13.0, macOS 10.15, *)
+public struct GIDSwiftUISignInButton: View {
```
Comment (petea) [inline GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift:24]: Do we want the `GID` prefix in the Swift environment?

Comment (petea) [inline GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift:24]: It looks like ObjC interop isn't possible for structs.  Are we OK with that?  Would it be possible to use a class here?

Comment (petea) [inline GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift:24]: It looks like ObjC interop isn't possible for structs.  Are we OK with that?  Would it be possible to use a class here?

File: GoogleSignInSwift/Sources/SignInButtonStyling.swift [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import SwiftUI
+
+#if os(iOS) || targetEnvironment(macCatalyst)
+import UIKit
```
Comment (petea) [inline GoogleSignInSwift/Sources/SignInButtonStyling.swift:20]: Where is this being used?

Comment (petea) [inline GoogleSignInSwift/Sources/SignInButtonStyling.swift:20]: Where is this being used?

File: GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift (hunk 18 above)
Comment (mdmathias) [reply GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift:24]: Do you have a use-case for this? We have an existing button working in ObjC/UIKit right now, and so using this SwiftUI button over there shouldn't be a need. Moreover, I believe SwiftUI enforces conformers of `View` to be value types. Using a `final class` (e.g., `final class MyView: View {...}`) will yield an error: "Fatal error: views must be value types (either a struct or an enum)."

If we want to expose this button to UIKit, then we'd need to use [`UIHostingController`](https://developer.apple.com/documentation/swiftui/uihostingcontroller) and [`NSHostingView`](https://developer.apple.com/documentation/swiftui/nshostingview).

Comment (mdmathias) [reply GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift:24]: Resolving as dupe.

File: GoogleSignInSwift/Sources/SignInButtonStyling.swift (hunk 19 above)
Comment (mdmathias) [reply GoogleSignInSwift/Sources/SignInButtonStyling.swift:20]: Resolving as a dupe.

Comment (mdmathias) [reply GoogleSignInSwift/Sources/SignInButtonStyling.swift:20]: Actually, this was silly. I don't believe these imports are needed here!

File: GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift (hunk 18 above)
Comment (mdmathias) [reply GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift:24]: What do you think about renaming to `GoogleSignInButton`

Comment (petea) [reply GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift]: Yes.  I think either `GoogleSignInButton` or `SignInButton`.

With namespacing:

```
GoogleSignInSwift.GoogleSignInButton
or
GoogleSignInSwift.SignInButton
```

and eventually:

```
GoogleSignIn.GoogleSignInButton
or
GoogleSignIn.SignInButton
```

Comment (petea) [reply GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift]: I'm thinking of the AppKit use case.  A traditional ObjC macOS app that wants to use the SwiftUI button.  It's an edge case, but it'd be nice to have a story.  One solution for the developer would be to write a AppKit view in Swift that uses our SwiftUI button via `NSHostingView`.

Comment (mdmathias) [reply GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift]: I went with `GoogleSignInButton`.

Comment (mdmathias) [reply GoogleSignInSwift/Sources/GIDSwiftUISignInButton.swift]: Tracking this with [a new issue](https://github.com/google/GoogleSignIn-iOS/issues/117).

**Review by petea — APPROVED** [review]
